File: examples/python/m2-tool-runtime/tools/bash.py
"""bash：参数化进程执行；默认关闭，也绝不经过 shell。"""

from __future__ import annotations

import subprocess
from typing import Any, Callable, Mapping, Sequence

from tool_types import ToolSpec
from workspace import Workspace

DEFAULT_TIMEOUT_S = 10.0
DEFAULT_MAX_OUTPUT_CHARS = 20_000
DEFAULT_ALLOWLIST = ("echo", "/bin/echo", "pwd", "/bin/pwd")
ProcessRunner = Callable[..., subprocess.CompletedProcess[str]]


class BashTool:
    """在 workspace cwd 下启动一个精确白名单程序。

    ``args`` 会逐项传给进程，不解释 ``;``、``|``、重定向或变量展开。
    allowlist、cwd、超时和输出截断只是教学级进程契约，不是 OS 沙箱。
    """

    def __init__(
        self,
        workspace: Workspace,
        *,
        enabled: bool = False,
        allowlist: Sequence[str] | None = None,
        default_timeout_s: float = DEFAULT_TIMEOUT_S,
        max_output_chars: int = DEFAULT_MAX_OUTPUT_CHARS,
        runner: ProcessRunner = subprocess.run,
    ) -> None:
        self.workspace = workspace
        self.enabled = enabled
        self.allowlist = tuple(allowlist) if allowlist is not None else DEFAULT_ALLOWLIST
        self.default_timeout_s = default_timeout_s
        self.max_output_chars = max_output_chars
        self._runner = runner
        self.spec = ToolSpec(
            name="bash",
            description="Run one allowlisted program in the workspace without a shell",
            input_schema={
                "type": "object",
                "required": ["program", "args"],
                "properties": {
                    "program": {"type": "string"},
                    "args": {"type": "array", "items": {"type": "string"}},
                    "timeout_s": {"type": "number"},
                },
                "additionalProperties": False,
            },
        )

    def validate_arguments(self, arguments: Mapping[str, Any]) -> None:
        """校验程序名、字符串参数数组和正超时。"""

        program = arguments.get("program")
        if not isinstance(program, str) or not program.strip():
            raise ValueError("program must be a non-empty string")
        args = arguments.get("args")
        if not isinstance(args, list) or any(not isinstance(item, str) for item in args):
            raise ValueError("args must be an array of strings")
        timeout = arguments.get("timeout_s")
        if timeout is not None and (
            not isinstance(timeout, (int, float))
            or isinstance(timeout, bool)
            or float(timeout) <= 0
        ):
            raise ValueError("timeout_s must be a positive number")

    def execute(self, arguments: Mapping[str, Any]) -> dict[str, Any]:
        """启动进程；非零退出和超时都抛错，由 Registry 收成结构化失败。"""

        if not self.enabled:
            raise RuntimeError("bash is disabled; enable explicitly for trusted demos only")

        program = str(arguments["program"])
        if program not in self.allowlist:
            raise RuntimeError("program rejected by exact allowlist")
        args = [str(item) for item in arguments["args"]]
        timeout = float(arguments.get("timeout_s") or self.default_timeout_s)
        try:
            completed = self._runner(
                [program, *args],
                shell=False,
                cwd=str(self.workspace.root),
                capture_output=True,
                text=True,
                timeout=timeout,
                check=False,
            )
        except subprocess.TimeoutExpired as error:
            raise RuntimeError(f"program timed out after {timeout}s") from error

        stdout, stdout_cut = self._truncate(completed.stdout or "")
        stderr, stderr_cut = self._truncate(completed.stderr or "")
        if completed.returncode != 0:
            detail = stderr.strip() or stdout.strip() or "no output"
            raise RuntimeError(f"exit {completed.returncode}: {detail}")
        return {
            "program": program,
            "args": args,
            "exit_code": completed.returncode,
            "stdout": stdout,
            "stderr": stderr,
            "truncated": stdout_cut or stderr_cut,
        }

    def _truncate(self, text: str) -> tuple[str, bool]:
        """限制单个输出流大小，避免将巨量日志塞回模型上下文。"""

        if len(text) <= self.max_output_chars:
            return text, False
        return text[: self.max_output_chars], True

File: examples/rust/m2-tool-runtime/src/tools/bash.rs
//! 参数化进程工具：不经过 shell，程序名按精确 allowlist 匹配。

use crate::registry::{require_object, require_string};
use crate::tool_types::{Tool, ToolError, ToolSpec};
use crate::workspace::Workspace;
use serde_json::{Value, json};
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};

const DEFAULT_TIMEOUT_S: f64 = 10.0;
const MAX_OUTPUT_CHARS: usize = 20_000;

/// 默认关闭的参数化进程工具。
pub struct BashTool {
    workspace: Workspace,
    enabled: bool,
    allowlist: Vec<String>,
    default_timeout_s: f64,
}

impl BashTool {
    /// 绑定 Workspace、启用开关和精确程序 allowlist。
    pub fn new(workspace: Workspace, enabled: bool, allowlist: Option<Vec<String>>) -> Self {
        Self {
            workspace,
            enabled,
            allowlist: allowlist.unwrap_or_else(|| {
                vec![
                    "echo".into(),
                    "/bin/echo".into(),
                    "pwd".into(),
                    "/bin/pwd".into(),
                ]
            }),
            default_timeout_s: DEFAULT_TIMEOUT_S,
        }
    }
}

impl Tool for BashTool {
    fn spec(&self) -> ToolSpec {
        ToolSpec {
            name: "bash".into(),
            description: "Run one allowlisted program in the workspace without a shell".into(),
            input_schema: json!({
                "type": "object",
                "required": ["program", "args"],
                "properties": {
                    "program": {"type": "string"},
                    "args": {"type": "array", "items": {"type": "string"}},
                    "timeout_s": {"type": "number"}
                },
                "additionalProperties": false
            }),
            strict: false,
        }
    }

    fn validate_arguments(&self, arguments: &Value) -> Result<(), ToolError> {
        let object = require_object(arguments)?;
        let _ = require_string(object, "program")?;
        let args = object
            .get("args")
            .and_then(Value::as_array)
            .ok_or_else(|| ToolError::new("args must be an array of strings"))?;
        if args.iter().any(|item| !item.is_string()) {
            return Err(ToolError::new("args must be an array of strings"));
        }
        if let Some(timeout) = object.get("timeout_s") {
            let value = timeout
                .as_f64()
                .ok_or_else(|| ToolError::new("timeout_s must be a positive number"))?;
            if value <= 0.0 {
                return Err(ToolError::new("timeout_s must be a positive number"));
            }
        }
        Ok(())
    }

    fn execute(&mut self, arguments: &Value) -> Result<Value, ToolError> {
        if !self.enabled {
            return Err(ToolError::new(
                "bash is disabled; enable explicitly for trusted demos only",
            ));
        }
        let object = require_object(arguments)?;
        let program = require_string(object, "program")?;
        if !self.allowlist.iter().any(|allowed| allowed == program) {
            return Err(ToolError::new("program rejected by exact allowlist"));
        }
        let args = object
            .get("args")
            .and_then(Value::as_array)
            .ok_or_else(|| ToolError::new("args must be an array of strings"))?
            .iter()
            .map(|value| value.as_str().unwrap_or_default())
            .collect::<Vec<_>>();
        let timeout_s = object
            .get("timeout_s")
            .and_then(Value::as_f64)
            .unwrap_or(self.default_timeout_s);

        let mut child = Command::new(program)
            .args(&args)
            .current_dir(self.workspace.root())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .map_err(|error| ToolError::new(error.to_string()))?;
        let deadline = Instant::now() + Duration::from_secs_f64(timeout_s);
        loop {
            match child.try_wait() {
                Ok(Some(_)) => break,
                Ok(None) if Instant::now() >= deadline => {
                    let _ = child.kill();
                    let _ = child.wait();
                    return Err(ToolError::new(format!(
                        "program timed out after {timeout_s}s"
                    )));
                }
                Ok(None) => std::thread::sleep(Duration::from_millis(10)),
                Err(error) => return Err(ToolError::new(error.to_string())),
            }
        }
        let output = child
            .wait_with_output()
            .map_err(|error| ToolError::new(error.to_string()))?;
        let (stdout, stdout_cut) =
            truncate_chars(String::from_utf8_lossy(&output.stdout).into_owned());
        let (stderr, stderr_cut) =
            truncate_chars(String::from_utf8_lossy(&output.stderr).into_owned());
        if !output.status.success() {
            let code = output.status.code().unwrap_or(-1);
            let detail = if stderr.trim().is_empty() {
                stdout.trim()
            } else {
                stderr.trim()
            };
            return Err(ToolError::new(format!(
                "exit {code}: {}",
                if detail.is_empty() {
                    "no output"
                } else {
                    detail
                }
            )));
        }
        Ok(json!({
            "program": program,
            "args": args,
            "exit_code": output.status.code().unwrap_or(0),
            "stdout": stdout,
            "stderr": stderr,
            "truncated": stdout_cut || stderr_cut,
        }))
    }
}

fn truncate_chars(text: String) -> (String, bool) {
    if text.chars().count() <= MAX_OUTPUT_CHARS {
        return (text, false);
    }
    (text.chars().take(MAX_OUTPUT_CHARS).collect(), true)
}
